mmv1/third_party/terraform/utils/privateca_utils.go

```diff
@@ -13,6 +14,234 @@ import (
 // deleted on the API-side. This flattener creates default objects for sub-objects that match this pattern
 // to fix perma-diffs on default-only objects. For this to work all objects that are flattened from nil to
 // their default object *MUST* be set in the user's config, so they are all marked as Required in the schema.
+//
+// This file also contains shared expanders between the above resources. The expanders are required in order
+// to handle the optional primitive field in CaOptions. By adding a virtual field, the expander can distinguish
+// between an unset primitive field and a set primitive field with a default value.
+
+// Expander utilities
+
+func expandPrivatecaCertificateConfigX509ConfigCaOptions(v interface{}, d TerraformResourceData, config *Config) (interface{}, error) {
+	// Virtual fields to distinguish between unset booleans and booleans set with a default value.
+	// Unset is_ca or unset max_issuer_path_length either allow any values for these fields when
+	// used in an issuance policy, or allow the API to use default values when used in a
+	// certificate config. A default value of is_ca=false means that issued certificates cannot
+	// be CA certificates. A default value of max_issuer_path_length=0 means that the CA cannot
+	// issue CA certificates.
+
+	// includeIsCa can be nil for Certificate Authority resources
+	includeIsCa := d.Get("include_is_ca")
+	includePathLength := d.Get("include_max_path_length")
+
+	if v == nil {
+		if includeIsCa != nil && includeIsCa.(bool) {
+			return nil, fmt.Errorf("include_is_ca cannot be set without setting ca_options.is_ca")
+		}
+		if includePathLength.(bool) {
+			return nil, fmt.Errorf("include_max_path_length cannot be set without setting ca_options.max_issuer_path_length")
+		}
+		return v, nil
+	}
+	l := v.([]interface{})
+	if len(l) == 0 || l[0] == nil {
+		if includeIsCa != nil && includeIsCa.(bool) {
+			return nil, fmt.Errorf("include_is_ca cannot be set without setting ca_options.is_ca")
+		}
+		if includePathLength.(bool) {
+			return nil, fmt.Errorf("include_max_path_length cannot be set without setting ca_options.max_issuer_path_length")
+		}
+		return nil, nil
+	}
+
+	raw := l[0]
+	original := raw.(map[string]interface{})
+	transformed := make(map[string]interface{})
+
+	//if original["is_ca"].(bool) && !includeIsCa.(bool) {
+	//	return nil, fmt.Errorf("You must set include_is_ca=true with ca_options.is_ca=true")
+	//}
+	//if original["max_issuer_path_length"].(int) > 0 && !includePathLength.(bool) {
+	//	return nil, fmt.Errorf("You must set include_max_path_length=true with ca_options.max_issuer_path_length>0")
+	//}
+
+	if (includeIsCa != nil && includeIsCa.(bool)) || original["is_ca"].(bool) {
```

Comment:
It feels weird to ignore the values of includeIsCa if isCa is set. It means that includeIsCa is only used when isCa is unset. Maybe we should enforce that includeIsCa is only false when isCa is false

Reply:
The latter half of the conditional was meant to resolve concerns around this being a breaking change. Ideally, we'd enforce that `includeIsCa` is set to use non-defaults, but we can't since that's breaking. I think it'd be confusing to describe the purpose of `includeIsCa` if it's only used to distinguish between unset and false.